File: daily/2026/2196_Create_Binary_Tree_From_Descriptions.cpp
#include <iostream>
#include <vector>
#include <unordered_set>
#include <queue>

using namespace std;


// Definition for a binary tree node.
struct TreeNode {
    int val;
    TreeNode *left;
    TreeNode *right;
    TreeNode() : val(0), left(nullptr), right(nullptr) {}
    TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
    TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
};

void printTree(TreeNode* root){
    queue<TreeNode*> que;
    que.push(root);
    while(!que.empty()){
        TreeNode* curr = que.front();
        que.pop();
        cout << curr->val << " ";
        if (curr->left) que.push(curr->left);
        if (curr->right) que.push(curr->right);
    }
    cout << endl;
    return;
}

class Solution {
public:
    TreeNode* createBinaryTree(vector<vector<int>>& descriptions) {
        unordered_map<int, TreeNode*> m;
        unordered_set<int> nodes;
        unordered_set<int> children;
        for (auto& t : descriptions){
            int p = t[0];
            int c = t[1];
            int l = t[2];
            nodes.insert(p);
            nodes.insert(c);
            children.insert(c);
            if (m.find(p) == m.end()){
                TreeNode* parent = new TreeNode(p);
                m[p] = parent;
            }
            if (m.find(c) == m.end()){
                TreeNode* child = new TreeNode(c);
                m[c] = child;
            }
            if (l == 1) m[p]->left = m[c];
            else m[p]->right = m[c];
        }
        for (int node : nodes) if (children.find(node) == children.end()) return m[node];
        return nullptr;
    }
};

int main(){
    vector<vector<int>> descriptions = {{20,15,1},{20,17,0},{50,20,1},{50,80,0},{80,19,1}};
    // vector<vector<int>> descriptions = {{1,2,1},{2,3,0},{3,4,1}};
    Solution S;
    TreeNode* root = S.createBinaryTree(descriptions);
    printTree(root);
    return 0;
}

File: daily_python/2196_Create_Binary_Tree_From_Descriptions.py
import os
import sys
from typing import List, Optional
from collections import deque

# Definition for a binary tree node.
class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right
    def printTree(self):
        print(self.val, end = ' ')
        que = deque([self.left, self.right])
        while que:
            curr = que.popleft()
            if curr:
                print(curr.val, end = ' ')
                que.append(curr.left)
                que.append(curr.right)
            else:
                print(-1, end = ' ')
        print()
        return

class Solution:
    def createBinaryTree(self, descriptions: List[List[int]]) -> Optional[TreeNode]:
        nodes = {}
        children = set()
        for p, c, l in descriptions:
            if p not in nodes:
                nodes[p] = TreeNode(p)
            if c not in nodes:
                nodes[c] = TreeNode(c)
            children.add(c)
            if l == 0:
                nodes[p].right = nodes[c]
            else:
                nodes[p].left = nodes[c]
        for val, node in nodes.items():
            if val not in children:
                return node
        return None
    
def readDataSet(filename):
    dataset = []
    with open(filename, 'r') as file:
        content = file.read().strip()
        blocks = content.split('\n\n')
        for block in blocks:
            lines = block.split('\n')
            descriptions = [list(map(int, description.split(','))) for description in lines[0].split('=')[1].strip()[2:-3].split('},{')]
            dataset.append(descriptions)
    return dataset

if __name__ == '__main__':
    if len(sys.argv) == 1:
        filename = os.path.splitext(__file__)[0] + '.txt'
    else:
        filename = sys.argv[1]
    dataset = readDataSet(filename)
    results = []
    solution = Solution()
    for descriptions in dataset:
        results.append(solution.createBinaryTree(descriptions))
    for index, result in enumerate(results):
        print(f'Example {index + 1} : ', end = '')
        result.printTree() if result else print(-1)